src/index.ts

```diff
@@ -256,13 +256,12 @@ const mkMatchXW =
   <C extends B[keyof B]>(x: A): C => {
     const tag = x[tagKey] as Tag<A>
 
-    const g = fs[tag]
     // eslint-disable-next-line functional/no-conditional-statement, @typescript-eslint/no-unsafe-return
-    if (g !== undefined) return g as C
+    if (Object.prototype.hasOwnProperty.call(fs, tag)) return fs[tag] as C
```

Comment:
Could we add a test that checks we're using `hasOwnProperty`?

Reply:
> dp you mean we should check that something like `toString` will not match anything?

Yeah, I was thinking something like this.

Reply:
13c9b3010f65adcba3ea93a806df8ecfe264753a